File: cses/06_tree_algorithms/1137_subtree-queries/solution.py
import sys


def main():
    data = sys.stdin.buffer.read().split()
    idx = 0
    n = int(data[idx]); idx += 1
    q = int(data[idx]); idx += 1

    val = [0] * (n + 1)
    for i in range(1, n + 1):
        val[i] = int(data[idx]); idx += 1

    # Danh sách kề của cây.
    adj = [[] for _ in range(n + 1)]
    for _ in range(n - 1):
        a = int(data[idx]); idx += 1
        b = int(data[idx]); idx += 1
        adj[a].append(b)
        adj[b].append(a)

    # DFS lặp từ gốc 1 để tính Euler tour: gán tin khi "vào" node, gán tout khi
    # "ra" (đã duyệt hết con). Subtree của node u ứng với đoạn [tin[u], tout[u]].
    tin = [0] * (n + 1)
    tout = [0] * (n + 1)
    parent = [0] * (n + 1)
    child_ptr = [0] * (n + 1)  # con trỏ tới con tiếp theo cần duyệt của mỗi node
    visited = [False] * (n + 1)

    order_stack = [1]
    visited[1] = True
    timer = 1
    tin[1] = timer
    while order_stack:
        u = order_stack[-1]
        if child_ptr[u] < len(adj[u]):
            v = adj[u][child_ptr[u]]
            child_ptr[u] += 1
            if not visited[v]:
                visited[v] = True
                parent[v] = u
                timer += 1
                tin[v] = timer  # gán tin cho con khi vừa đi vào
                order_stack.append(v)
        else:
            tout[u] = timer  # gán tout khi mọi con đã được thăm
            order_stack.pop()

    # BIT (Fenwick tree) 1-indexed trên miền [1, n] theo tin.
    bit = [0] * (n + 1)

    def update(i, delta):
        while i <= n:
            bit[i] += delta
            i += i & (-i)

    def prefix(i):
        s = 0
        while i > 0:
            s += bit[i]
            i -= i & (-i)
        return s

    # Khởi tạo BIT: đặt giá trị của mỗi node tại vị trí tin.
    for i in range(1, n + 1):
        update(tin[i], val[i])

    out = []
    for _ in range(q):
        t = data[idx]; idx += 1
        if t == b'1':
            # Đổi giá trị node s: cập nhật điểm với delta = giá trị mới - giá trị cũ.
            s = int(data[idx]); idx += 1
            x = int(data[idx]); idx += 1
            update(tin[s], x - val[s])
            val[s] = x
        else:
            # Tổng subtree của s = tổng đoạn [tin[s], tout[s]].
            s = int(data[idx]); idx += 1
            out.append(prefix(tout[s]) - prefix(tin[s] - 1))

    sys.stdout.buffer.write(("\n".join(map(str, out)) + ("\n" if out else "")).encode())


main()

File: cses/06_tree_algorithms/1137_subtree-queries/solution.cpp
#include <iostream>
#include <vector>

using namespace std;

// BIT (Fenwick tree) 1-indexed: hỗ trợ point update và range sum trong O(log n).
class FenwickTree {
public:
    explicit FenwickTree(int n) : bit_(n + 1) {}

    void add(int index, long long delta) {
        for (; index < static_cast<int>(bit_.size()); index += index & -index) {
            bit_[index] += delta;
        }
    }

    long long prefix_sum(int index) const {
        long long result = 0;
        for (; index > 0; index -= index & -index) {
            result += bit_[index];
        }
        return result;
    }

    long long range_sum(int left, int right) const {
        return prefix_sum(right) - prefix_sum(left - 1);
    }

private:
    vector<long long> bit_;
};

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    int n, query_count;
    cin >> n >> query_count;
    vector<long long> value(n + 1);
    for (int node = 1; node <= n; ++node) {
        cin >> value[node];
    }

    // Danh sách kề của cây.
    vector<vector<int>> graph(n + 1);
    for (int i = 0; i < n - 1; ++i) {
        int a, b;
        cin >> a >> b;
        graph[a].push_back(b);
        graph[b].push_back(a);
    }

    // Euler tour bằng DFS lặp từ gốc 1: gán entry (tin) khi vào node, exit (tout)
    // khi đã duyệt hết con. Subtree của node u ứng với đoạn [entry[u], exit[u]].
    vector<int> parent(n + 1, -1), child_index(n + 1);
    vector<int> entry(n + 1), exit(n + 1);
    vector<int> stack = {1};
    parent[1] = 0;
    int timer = 1;
    entry[1] = timer;
    while (!stack.empty()) {
        int node = stack.back();
        if (child_index[node] < static_cast<int>(graph[node].size())) {
            int next = graph[node][child_index[node]++];
            if (next == parent[node]) {
                continue;  // bỏ qua cạnh quay ngược về cha
            }
            parent[next] = node;
            entry[next] = ++timer;  // gán tin cho con khi vừa đi vào
            stack.push_back(next);
        } else {
            exit[node] = timer;  // gán tout khi mọi con đã được thăm
            stack.pop_back();
        }
    }

    // Khởi tạo BIT: đặt giá trị của mỗi node tại vị trí entry[node].
    FenwickTree fenwick(n);
    for (int node = 1; node <= n; ++node) {
        fenwick.add(entry[node], value[node]);
    }

    while (query_count--) {
        int type, node;
        cin >> type >> node;
        if (type == 1) {
            // Đổi giá trị node: cập nhật điểm với delta = giá trị mới - giá trị cũ.
            long long new_value;
            cin >> new_value;
            fenwick.add(entry[node], new_value - value[node]);
            value[node] = new_value;
        } else {
            // Tổng subtree = tổng đoạn [entry[node], exit[node]].
            cout << fenwick.range_sum(entry[node], exit[node]) << '\n';
        }
    }
}
